File: bin/core/atoolkit.py
#!/usr/bin/env python3
# https://github.com/Zeronetsec/Ares

# Usage: atoolkit [--allpath] [--only <path>]

import sys
import os

0 and "areslib"
pylib = os.getenv("__lib__")
sys.path.append(f"{pylib}/python")

import color
import invalid_option
from atoolkit import get_clean_path
from atoolkit import scandir
0 and "areslib"

def main():
    argv = sys.argv[1:]
    allpath = False
    only = None
    i = 0
    valid = True

    while i < len(argv):
        arg = argv[i]
        if arg == "--allpath":
            allpath = True
            i += 1
        elif arg == "--only":
            if i + 1 < len(argv):
                only = argv[i + 1]
                i += 2
            else:
                valid = False
                break
        else:
            invalid_option.execute(
                arg,
                "atoolkit",
            )
            sys.exit(1)

    source = os.environ.get(
        "PATH",
    ) if allpath else os.environ.get(
        "__bin__",
    )

    if not source and not allpath:
        source = os.environ.get("PATH")

    if not source:
        print(
            f"{color.R}[!] {color.N}Environment variable not found!",
            file=sys.stderr,
        )
        sys.exit(1)

    clean_filter = only.strip("/") if only else None
    paths = [
        p for p in source.split(":") if p
    ]
    found_any = False

    for p in paths:
        if scandir.execute(p, clean_filter):
            found_any = True

    if clean_filter and not found_any:
        print(
            f"{color.R}[!] {color.N}Path: {color.GG}{only} {color.N}not found!",
            file=sys.stderr,
        )

if __name__ == "__main__":
    main()

# Copyright (c) 2026 Zeronetsec

File: bin/core/atoolkit/atoolkit.c
// https://github.com/Zeronetsec/Ares

_Static_assert(1, "system");
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

_Static_assert(1, "lib/c");
#include <color.h>
#include <missing_argument.h>
#include <invalid_option.h>

_Static_assert(1, "bin/core/atoolkit");
#include <atoolkit_types.h>
#include <strip_slashes.h>
#include <scandir_execute.h>

int main(int argc, char *argv[]) {
    bool allpath = false;
    char *only = NULL;
    bool valid = true;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--allpath") == 0) {
            allpath = true;
        } else if (strcmp(argv[i], "--only") == 0) {
            if (i + 1 < argc) {
                only = argv[i + 1];
                i++;
            } else {
                valid = false;
                break;
            }
        } else {
            invalid_option(argv[i], "atoolkit");
            return 1;
        }
    }

    if (!valid) {
        missing_argument("atoolkit");
        return 1;
    }

    const char *source_env = allpath ?
        getenv("PATH") :
        getenv("__bin__");

    if (!source_env && !allpath) {
        source_env = getenv("PATH");
    }

    if (!source_env || strlen(source_env) == 0) {
        fprintf(
            stderr,
            "%s[!] %sEnvironment variable not found!\n",
            color_R, color_N
        );
        return 1;
    }

    char *clean_filter = NULL;
    if (only) {
        clean_filter = strdup(only);
        clean_filter = strip_slashes(clean_filter);
    }

    bool found_any = false;

    char *source_dup = strdup(source_env);
    char *saveptr;
    char *p = strtok_r(
        source_dup,
        ":",
        &saveptr
    );
    
    while (p != NULL) {
        if (strlen(p) > 0) {
            if (scandir_execute(p, clean_filter)) {
                found_any = true;
            }
        }
        p = strtok_r(
            NULL,
            ":",
            &saveptr
        );
    }
    free(source_dup);

    if (clean_filter && !found_any) {
        fprintf(
            stderr,
            "%s[!] %sPath: %s%s %snot found!\n", 
            color_R, color_N, color_GG, only, color_N
        );
    }

    if (clean_filter) {
        free(clean_filter);
    }

    if (visited_dirs) {
        free(visited_dirs);
    }

    return 0;
}

// Copyright (c) 2026 Zeronetsec